### ISSUE TITLE
ExcelDna.Loader.XlAddIn.Initialize call failed (Again)

### ISSUE THREAD

**Issue by TerrySnow1963** [Excel-DNA/ExcelDna#205]
This issue is a repeat of issue #75 , with some specific information.
I am using ExcelDNA version 0.34.6373.42344, and I have produced a packed DLL.

I am testing the XLL in 2 platforms: Windows 7 64-Bit with 32-bit Excel 2013, and Windows 10 64-bit with 32-bit Excel. The Windows 7 machine is also the machine where my development environment is located.

Everything is fine for the Windows 7 machine, but for the Windows 10 Machine, I mistakenly added in the 64-bit packed version first, I then added in the 32-bit packed version, and started getting the ExcelDna.LoaderXLAddIn.Initialize call fail error. I removed all XLLs from the Windows10 machine, then just added in the 32-bit packed addin and added it to Excel, but I continue to get the ExcelDna.Loader.XlAddIn.Initialize call failure. I have searched the registry and the PC, restarted the PC, but I still keep getting the error.

Do you have any suggestions to resolve this ?

**Comment by TerrySnow1963** [comment]
I should also mentioned that my development environment is Visual Studio 2015

**Comment by TerrySnow1963** [comment]
I have done some more investigation. I have repeated this process using a new ExcelDNA library using the sample code from your site. This time I just added in the 32-bit packed XLL (on the Windows 10 machine). 
I still get the same type of error (but referring to the new Packed Library).

Some other observations. I get the error dialog twice in succession when I try to deselect the addin from Excel Addin manager

**Comment by govert** [comment]
You might have a mismatched version of the ExcelDna.Integration.dll assembly, either in the same directory as the .xll, or in the .NET GAC.

**Comment by TerrySnow1963** [comment]
Not sure how I could have got a mismatched version of ExcelDNA.Integeration.dll, I have only ever downloaded 1 version of ExcelDNA (using NuGet), unless it was about mistake I made adding in the 64-bit packed XLL. There is definitely not a copy of the integration DLL in the local directory. I will investigate how to inspect the GAC to confirm.

The thing that throws me is that I have an installation of the same addin on 2 different OS (A windows 7  machine and a windows 10 machine) and literally this was a copy of the same packed XLL to both.

I think that the error that I am getting from the loader is misleading.

**Comment by govert** [comment]
I'm not sure what you mean by "the same type of error". It would help to see a screenshot or at least the exact text of the error. The message that is described in #75 has nothing to do with 64-bit / 32-bit, but might result from errors when the add-in is loaded, e.g. if there are other dependencies that are not resolved correctly.

Excel sometimes wants to copy the add-in into its own directory - this can cause confusion or trouble if your add-in needs other files than just the .xll to work.

**Comment by TerrySnow1963** [comment]
The error I get is specifically the same error (same dialog box as seen in the screenshot) that was reported in #75 (with the only difference being the name of the source adding). 

I only have 1 DLL (Vb Class Library) that I have an ExcelDNA wrapper around, and this is bundled in to the packed DLL.

I have done a full machine search for ExcelDNA.Integeration.dll and I am pretty sure that it does not exist as a standalone DLL in some excel directory (or any directory on the machine)
I will look in the GAC, but I don't think this will be the solution, as I have rebooted the PC to clear the memory.

**Comment by govert** [comment]
You can check the GAC by going in the Windows File Explorer:
* C:\Windows\Microsoft.NET\assembly\GAC_MSIL
But that seems unlikely to be the problem.

Otherwise I suggest these steps:

* Download the full Excel-DNA distribution .zip file from here : https://github.com/Excel-DNA/ExcelDna/releases/download/v0.34.6/ExcelDna-0.34.6.zip

* Unzip to a convenient location

* Attempt to open `Distribution\ExcelDna.xll` in Excel

* You should get a function called `=AddThem(...)`

That tests a simple add-in, and gives an indication of whether the problem is related to your add-in at all.

**Comment by TerrySnow1963** [comment]
Thanks , I will try tomorrow morning and report back

**Comment by TerrySnow1963** [comment]
I have checked the GAC, looked for ExcelDNA.Integeration.dll, under its own folder (no such folder), or anything under Excel like folders as well as doing File Explorer search. The dll is not in the GAC.

As per your suggestion, I have downloaded the Excel-DNA distribution, unzipped it, and added it into Excel. The ExcelDNA.xll addin worked fine.

So there is something wrong with my addin, and what ever is wrong, the packed XLL will work on Windows 7 Excel 2013 32-bit machines (has been tested on 2 separate W7 machines) and does not work on Windows 10 Excel 2013 32-bit machines (has been tested on 2 separate W10 machines).

Looking at my DNA file, perhaps I should not have the keywords for ExplicitExports and LoadFromBytes set in the Reference Path entry (?), I will try removing that.

Here is my DNA file from the project.
<DnaLibrary Name="RTAFuncLibXLExt Add-In" RuntimeVersion="v4.0">
  <ExternalLibrary Path="RTAFuncLibXLExt.dll" ExplicitExports="false" LoadFromBytes="true" Pack="true" />
  <Reference Path="RTAFuncLibExt.dll" ExplicitExports="false" LoadFromBytes="true" Pack="true" />
  

</DnaLibrary>

**Comment by TerrySnow1963** [comment]
Sorry, DNA file should not have been in XML format

![image](https://user-images.githubusercontent.com/17805024/41874576-a0443a54-790b-11e8-970e-82fa667f97f3.png)

**Comment by govert** [comment]
What other libraries do you reference from RTAFuncLibXlExt.dll and from RTAFbcLibExt.dll? Maybe one of these are installed as part of the system or otherwise on the Windows 7 machines, and not Windows 10.
There is a tool called the 'Fusion Logger' that can help you to find assemblies that .NET tries to load but fails.
Sometimes the output when running in a debugger can also show what assemblies fail to load.

**Comment by TerrySnow1963** [comment]
I will check out Fusion Logger. In meantime here is screenshot of VS Solution Explorer showing references.
Rest of references are standard dot net
![image](https://user-images.githubusercontent.com/17805024/41875534-7662ed5e-790e-11e8-9822-52abf1cd6f0a.png)

**Comment by govert** [comment]
I'm sure it's just something small. If you're willing to email me your project, I'm happy to have a look.

**Comment by TerrySnow1963** [comment]
Thanks for offer, I think the issue is to do with the SOE (Standard Operating Environment) of the Windows 10 machines, so I dont think that you are going to be able to replicate the problem. The Windows 10 machines are corporate managed PCs/Laptops for which I will need to arrange Admin access to run Fusion Logger.

As an additional test, I installed the addin on a Windows 2012 Server running Excel 2010, where I have an admin account. The addin worked fine.

My steps from here: Get Admin privileges on my corporate Windows 10 Laptop, and run Fusion Logger to get more detail on assembly that is missing.

As I mentioned earlier, I suspect that the error reported by ExcelDNA is misleading.

**Comment by govert** [comment]
I think you're on the right track with the packing being the issue, and maybe shouldn't bother with the fusion stuff for now.

I remember some security software than caused trouble. Can you try the ExcelDna.xll and ExcelDna.dna files alone in a directory? Then it will be forced to use the packed ExcelDna.Integration.dll, and not the one from the file.
Also, going against my previous advice, can you try your add-in (the unpacked one), but adding in the ExcelDna.Integration.dll from you Nuget pavkage directory (or set Copy Local=true on the reference to have a rebuild put it there).

**Comment by garethhayter** [comment]
Can you use ILMerge or a similar tool to merge your assemblies into a single assembly which you then pack? I've only experienced an unpacking problem when there are external assemblies - this approach would ensure that never happens.

**Comment by TerrySnow1963** [comment]
@govert , I will try both recommendations and get back.
@garethhayter , I have not used ILMerge before, I will research it, I am not sure what the DNA file would look like, do you have an example where you have merged and then packed ?

Just confirming, I assume that I do not have to explicitly reference ExcelDna.Integration.dll in the DNA file. I assume that ExcelDna.Integration.dll is packed by default ?. The reason that I say this is that the packed dll works fine (with no other files) on fresh Windows7 machines.

The other step I am taking is getting a hold of a Windows10 machine that is not corporate managed (not using corporate security software).

Also I considered the Excel Options Trust Centre, and added my addin directory to the trusted directories, but that did not change the error.

**Comment by garethhayter** [comment]
@TerrySnow1963 , I use SmartAssembly to merge assemblies (as well as to do other things). ILMerge performs the same function with regards to merging. I haven't used ILMerge for many years, but I think it should still do the job of merging multiple assemblies into one: [ILMerge](https://github.com/Microsoft/ILMerge).

**Comment by TerrySnow1963** [comment]
@govert Feedback on your first recommendation. I placed the ExcelDNA.XLL and ExcelDna.dna files in separate folder, and tried to add the ExcelDna.XLL as addin to excel - I got the same error (XcelDna.xll addin loader: The ExcelDnaLoader.XLAddin.Initialize call failed.

**Comment by govert** [comment]
@TerrySnow1963  OK - so the next step is to change the reference that your project has to `ExcelDna.Integration.dll` to be "Copy Local: True". This will put it in the output directory. Then you might have a working (unpacked) add-in comprising the files:
* RTAFuncLibXlExt-AddIn.xll
* RTAFuncLibXlExt-AddIn.dna
* RTAFuncLibXlExt.dll
* RTAFuncLibExt.dll
* ExcelDna.Integration.dll

(For 64-bit Excel you'd need the AddIn64 versions of the first two files.)

If that doesn't work, you can further also edit the .dna file and set `LoadFromBytes="false"` in the two places it occurs.

**Comment by TerrySnow1963** [comment]
@govert Thanks, I will test this approach. First, I want to check the packed XLL on a 'clean' Windows10 machine. I am having one built and I will have access to the machine this morning.

I would like to stay with the Packed XLL approach, if possible, as it simplifies distribution. 

I would like to get to the bottom of why the packed XLL will work on the Windows7 and WindowsServer2012 machines and not the Windows10 machine.

**Comment by govert** [comment]
Are any anti-virus or other security tools loaded on those machines?
This does not sound like normal Windows security setting that are preventing the packed add-ins from running.
Sometimes one can get a clue by looking at the list of all .dlls that are loaded into the Excel process. You'd be looking for some anti-virus or other security software that is able to interfere with some of the API calls inside Excel.

Are you able to load a debugger or Visual Studio onto one of the problematic machines?

**Comment by govert** [comment]
Have you been able to try the unpacked add-in with the ExcelDna.Integration.dll on your machine?

On 28 Jun 2018, at 01:30, TerrySnow1963 <notifications@github.com<mailto:notifications@github.com>> wrote:


Some more info.
I have acquired a fresh, clean Windows10 machine (Windows Surface Pro Laptop running Windows 10 Pro 64-bit, with Excel 2013 32-bit)
The packed XLL executes fine on this clean machine.

Some info on the problem machines, these are all machines with the same corporate SOE, the corporate machines all run securiry / antivirus through the Windows Defender Security Center.

I will look at getting Visual Studio Installed on my corporate laptop

**Comment by TerrySnow1963** [comment]
No - I have not tried to generate an unpacked add-in yet. 
Today I wanted to test the clean Windows10 Machine, which showed me that a clean machine can run the packed DLL, so there is something about the corporate security on Windows 10 that is stopping the packed XLL.  You have given me an unpacked XLL previously which could run successfully, so I expect the same will happen for my XLL when it is unpacked. I would rather have it packed, so I will get local admin access to the corporate laptop and install Visual Studio and check out what is stopping the unpacking. I am waiting for the IT department to process the request for local admin access.

When I get to that point of debugging the admin load, I might get some advice on where to set breakpoints in the ExcelDNA code.

**Comment by govert** [comment]
@TerrySnow1963 You might as well email me directly. We can post something on the final resolution back here.

**Comment by TerrySnow1963** [comment]
Sorry, it has been a while, but the project has come back to life. Some changes were made to some of the functions, so I made a new version. I still had the problem of Windows 10 failing to unpack the packed version.

I took goverts advice and produced an unpacked version of the XLL, and the Addin can now be loaded into our corporate Windows 10 PCs.

I still have not determined why the xll load fails for a packed dll on our corporate Windows 10 PCs, but I am prepared to live with this as is.
